staging/src/k8s.io/api/bulk/v1alpha1/types.go

```diff
@@ -0,0 +1,216 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +genclient:nonNamespaced
+
+// BulkRequest contains a single request to bulk api.
+type BulkRequest struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Opaque identifier used to match requests and responses.
+	// +optional
+	RequestID string `json:"requestId,omitempty" protobuf:"bytes,1,opt,name=requestId"`
+
+	// Starts new watch.
+	// +optional
+	Watch *WatchOperation `json:"watch,omitempty" protobuf:"bytes,2,opt,name=watch"`
+
+	// Starts new watch.
+	// +optional
+	WatchList *WatchListOperation `json:"watchList,omitempty" protobuf:"bytes,3,opt,name=watchList"`
```

Comment:
Why do we need separate high-level "Watch" and "WatchList"?

Looking into types below, I would prefer merging those two and making "Name" field simply an optional.

Reply:
Not all options are valid/meaningful for single-item watch.
For example current implementation of watch (not bulk) does not allow 'Name' and 'FieldSelector' at the same time.
Also LabelSelector/Limit/Contine fields doesn't not make sense for single-watch. 

I'd like to make the new api "type-safe" as much of possible.
It is usually better not to add a field at all instead of runtime checks "such combination is not allowed".
In our case with two different messages user just can't provide unsupported combinations of fields (there is an exception for ListOptions.Watch field, it is just ignored).

PS: Actually at the beginning I used single 'verb' for both single- and list-watches.
After splitting the implementation also became a bit more structured and readable.

Reply:
Actually, there isn't anything really protecting you from passing label/field selector for single-item watch (non-bulk). 

And there are already fields in "ListOptions" that doesn't make sense for the watch case (limit or continue).
So I would really like to avoid splitting those two.

Reply:
All watch is implemented as WatchList.  The singular watch is just a convenience.  This API should reflect regular WatchList, not the convenience function.

Reply:
I'm not seeing from this why you need anything other than ListOptions.  Everything you are doing is covered by either field selectors or the existing options.

Reply:
> I'm not seeing from this why you need anything other than ListOptions. Everything you are doing is covered by either field selectors or the existing options.

We want to make sure this is able to authorize watches of individual items by name, to allow kubelets to watch individual secrets. If you use field selectors for that, that means special treatment of the metadata.name selector in the authorization check. Not sure if that’s good or bad, just calling it out

Reply:
@smarterclayton - I think you need to be able to specify group-version-kind. So we need something on top of ListOptions.

About @liggitt `s point - I think it a very good point to keep in mind.
I think I would still prefer to not introduce two separate Watch and WatchList operations (we already have some logic in code that is treating "metadata.name" field specially, e.g.:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/selection_predicate.go#L127
), but let's make conscious decision here.